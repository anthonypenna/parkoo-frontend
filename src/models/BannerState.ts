import { BannerType } from '@/constants/banner'

export interface BannerState {
  text: string
  type: BannerType
  visible: boolean
}
