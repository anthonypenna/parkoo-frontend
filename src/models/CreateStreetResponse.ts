import { Street } from './Street'

export interface CreateStreetResponse {
  street: Street
}
