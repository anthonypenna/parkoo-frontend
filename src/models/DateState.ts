import { Day } from '@/constants/date'

export interface DateState {
  currentDay: Day
}
