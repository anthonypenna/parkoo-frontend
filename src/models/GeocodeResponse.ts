import { StreetData } from './StreetData'

export interface GeocodeResponse {
  features: StreetData[]
}
