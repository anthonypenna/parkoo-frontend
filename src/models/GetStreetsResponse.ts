import { Street } from './Street'

export interface GetStreetsResponse {
  streets: Street[]
}
