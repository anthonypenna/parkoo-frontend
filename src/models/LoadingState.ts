export interface LoadingState {
  showLoading: boolean
}
