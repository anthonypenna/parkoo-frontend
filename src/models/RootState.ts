export interface RootState {
  showLoading: boolean
  hasFetchedGeolocation: boolean
  hasFetchedStreets: boolean
  showNoStreetsModal: boolean
  showAddStreetModal: boolean
}
