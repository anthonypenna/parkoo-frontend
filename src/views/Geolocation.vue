<template>
  <GeolocationErrorModal v-show="!showLoading" />
</template>

<script lang="ts">
import Vue from 'vue'
import GeolocationErrorModal from '@/components/molecules/GeolocationErrorModal.vue'
import { mapActions, mapGetters, mapMutations } from 'vuex'

export default Vue.extend({
  name: 'Geolocation',
  components: { GeolocationErrorModal },
  computed: {
    ...mapGetters(['showLoading'])
  },
  methods: {
    ...mapActions('user', ['getPosition']),
    ...mapMutations(['setShowLoading', 'setHasFetchedGeolocation'])
  },
  mounted() {
    this.getPosition()
      .then(() => {
        this.setHasFetchedGeolocation(true)
        this.$router.push({ name: 'Streets' })
      })
      .catch(() => {
        this.setShowLoading(false)
      })
  }
})
</script>
