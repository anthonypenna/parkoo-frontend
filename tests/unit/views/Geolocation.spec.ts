import { RootState } from '@/models/RootState'
import { createLocalVue, shallowMount, ThisTypedShallowMountOptions } from '@vue/test-utils'
import { VueConstructor } from 'vue'
import VueRouter from 'vue-router'
import Vuex, { Store } from 'vuex'
import Geolocation from '@/views/Geolocation.vue'
import { promisify } from 'util'

const $nextTick = promisify(process.nextTick)

const routerConfig = {
  routes: [
    { path: '/geolocation', name: 'Geolocation' },
    { path: '/streets', name: 'Streets' }
  ]
}

const storeConfig = {
  modules: {
    loading: {
      namespaced: true,
      actions: {
        stopLoading: jest.fn()
      },
      getters: {
        showLoading: jest.fn(() => true)
      },
      mutations: {
        setShowLoading: jest.fn()
      }
    },
    user: {
      namespaced: true,
      actions: {
        getPosition: jest.fn()
      },
      mutations: {
        setHasFetchedGeolocation: jest.fn()
      }
    }
  }
}

describe('<Geolocation />', () => {
  let localVue: VueConstructor<Vue>
  let router: VueRouter
  let store: Store<RootState>
  let componentOptions: ThisTypedShallowMountOptions<Vue>

  beforeEach(() => {
    localVue = createLocalVue()
    localVue.use(VueRouter)
    localVue.use(Vuex)
    router = new VueRouter(routerConfig)
    store = new Store(storeConfig)
    componentOptions = { localVue, router, store }
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should render correctly', () => {
    const wrapper = shallowMount(Geolocation, componentOptions)
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('when mounted', () => {
    beforeEach(() => {
      router.replace({ name: 'Geolocation' })
    })

    it('should geolocate the user', () => {
      shallowMount(Geolocation, componentOptions)
      expect(storeConfig.modules.user.actions.getPosition).toHaveBeenCalled()
    })

    describe('when geolocation fails', () => {
      beforeEach(() => {
        storeConfig.modules.user.actions.getPosition.mockRejectedValue({})
      })

      it('should stop loading', async () => {
        const wrapper = shallowMount(Geolocation, componentOptions)
        const setShowLoading = jest.spyOn<any, any>(wrapper.vm, 'stopLoading')
        await $nextTick()
        expect(setShowLoading).toHaveBeenCalled()
      })
    })

    describe('when geolocation succeeds', () => {
      beforeEach(() => {
        storeConfig.modules.user.actions.getPosition.mockResolvedValue({})
      })

      it('should set the fetched geolocation state', async () => {
        const wrapper = shallowMount(Geolocation, componentOptions)
        const setHasFetchedGeolocation = jest.spyOn<any, any>(wrapper.vm, 'setHasFetchedGeolocation')
        await $nextTick()
        expect(setHasFetchedGeolocation).toHaveBeenCalledWith(true)
      })

      it('should go to the streets page', async () => {
        shallowMount(Geolocation, componentOptions)
        const push = jest.spyOn(router, 'push')
        await $nextTick()
        expect(push).toHaveBeenCalledWith({ name: 'Streets' })
      })
    })
  })
})
