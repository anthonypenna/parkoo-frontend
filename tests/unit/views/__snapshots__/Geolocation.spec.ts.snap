// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Geolocation /> should render correctly 1`] = `<geolocationerrormodal-stub></geolocationerrormodal-stub>`;
